<a href="https://colab.research.google.com/github/chaegeon/Multi-learning/blob/main/%ED%8C%90%EB%8B%A4%EC%8A%A4Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# 데이터 분석용 필수 패키지
# 사용하지 않더라도 일단 무조건 import로 가져오고 보자
import numpy as np
import pandas as pd
# 넘파이가 암묵적으로 np라고 하듯이,
# 판다스도 암묵적으로 pd로 줄여 씀

# 판다스
- 파이썬에서 사용할 수 있는 데이터 전처리용 패키지
  - 엑셀처럼 파일을 다룰 수 있게 해주는 도구
  - 엑셀과의 가장 큰 차이점은 아주 큰 파일도 처리가 가능
    - 일반적으로 엑셀파일의 크기가 100MB만 넘어가도 잘 안 열릴 것
    - 판다스는 1GB가 넘어가는 아주 큰 파일도 빠르게 처리가 가능
    - 엑셀에서 간보고 가능하면 판다스로 넘어오는 등의 방식으로 처리한다고 하넹
  - 엑셀보다 더 복잡한 처리, 디비와 연동 등의 기능을 제공

- 데이터 분석
  - 분석 업무의 80%는 전처리 이다. 나머지 20%는 분석
    - 전처리 : 정제한다?
    - 잘 정제되지 않은 데이터를 가지고는 아무리 분석을 잘 해도 의미가 없다

## Series
  - 1차원 구조를 표현

In [ ]:
# 왼쪽에 보이는 건 인덱스
# 인덱스와 같이 표현돼서 2차원처럼 보이지만 1차원임
series = pd.Series( [10, 20, 30, 40] ) # series 라는 클래스
series

0    10
1    20
2    30
3    40
dtype: int64

In [ ]:
display( series.ndim )
display( series.shape )

1

(4,)

## DataFrame
- 2차원 구조
- 여러개의 시리즈가 모여서 하나의 데이터 프레임이 된다

- 넘파이 자료들이 모여서 하나의 시리즈가 되고
- 그 시리즈 들이 모여서 하나의 데이터프레임이 된다고 이해해도 될 듯
  - 호환이 된다

In [ ]:
weight = pd.DataFrame([
  [76.4, 'kg'],
  [75.7, 'kg'],
  [76, 'kg'],
  [76.2, 'kg']
])
weight

0   1
0  76.4  kg
1  75.7  kg
2  76.0  kg
3  76.2  kg

In [ ]:
display( weight. ndim)
display( weight. shape)

2

(4, 2)

## 파일 읽어오기
- 판다스는 'read_csv'라는 기능으로 파일을 읽어올 수 있다.
  - CSV(comma Seperated Value)
    - 텍스트가 콤마로 구분된 파일
  - 엑셀, csv,... 등의 다양한 파일 형태를 읽어서 **데이터프레임** 형태로 자동으로 변환을 해준다


- 데이터 폴더에서 공유받은 weight 파일 경로복사


In [ ]:
rawData = pd.read_csv('/content/drive/MyDrive/멀티캠퍼스/Data폴더/weight_log.csv')
rawData
# NaN은 결측치. 자료가 없는 부분 (넘파이 때 배웠음)

회차   이름         측정일   몸무게  단위   담당   지점
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구

In [ ]:
# 데이터 프레임 사용하기
 - 판다스는 기능이 너무 다양해서 일단은 기본적으로 사용하는 기능들만 다뤄보자

# 데이터 프레임 사용하기

## 출력
- head
  - 자료의 제일 앞 5개 출력
    - 디폴트 파라미터가 5로 지정되어 있음.
    - ()안에 파라미터 값을 지정해주면 됨
```
DataFrame.head(n=5)
```

- tail
  - 자료의 맨 끝 5개 출력
```
DataFrame.tail(n=5)
```

In [ ]:
# 상위 5개의 자료만 출력
rawData.head()

회차   이름         측정일   몸무게  단위   담당   지점
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구

In [ ]:
# 하위 5개의 자료만 출력
rawData.tail()

회차   이름         측정일   몸무게  단위   담당   지점
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구

In [ ]:
# head나 tail이나 파라미터의 기본값이 5로 지정되어 있음
# 괄호안에 파라미터를 지정해서 사용 가능
rawData.head(2)

회차   이름         측정일   몸무게  단위   담당   지점
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구

In [ ]:
rawData.tail(2)

회차   이름         측정일   몸무게  단위   담당   지점
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구

## 데이터 요약된 정보

In [ ]:
rawData.info()
# 몸무게에 6 non-null => 몸무게에 3개의 결치가 있다는 뜻
# object 는 문자열이라는 의미. 스트링을 안 쓰고 오브젝트는 쓴다고 함

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 7 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   회차      9 non-null      int64  
 1   이름      9 non-null      object 
 2   측정일     9 non-null      object 
 3   몸무게     6 non-null      float64
 4   단위      9 non-null      object 
 5   담당      9 non-null      object 
 6   지점      9 non-null      object 
dtypes: float64(1), int64(1), object(5)
memory usage: 632.0+ bytes


In [ ]:
# 자료의 기본적인 통계치
# 숫자 형태의 통계치.
# 이름 등의 문자값은 안 나옴
rawData.describe()

회차        몸무게
count  9.000000   6.000000
mean   5.000000  75.833333
std    2.738613   0.492612
min    1.000000  75.000000
25%    3.000000  75.700000
50%    5.000000  75.850000
75%    7.000000  76.150000
max    9.000000  76.400000

In [ ]:
# 문자값까지 통계치를 보고 싶다면
rawData.describe(include='all')
# top 은 빈도수

회차   이름         측정일        몸무게   단위   담당   지점
count   9.000000    9           9   6.000000    9    9    9
unique       NaN    1           9        NaN    1    3    4
top          NaN  홍길동  2020-03-04        NaN   kg  김현경  서초구
freq         NaN    9           1        NaN    9    4    4
mean    5.000000  NaN         NaN  75.833333  NaN  NaN  NaN
std     2.738613  NaN         NaN   0.492612  NaN  NaN  NaN
min     1.000000  NaN         NaN  75.000000  NaN  NaN  NaN
25%     3.000000  NaN         NaN  75.700000  NaN  NaN  NaN
50%     5.000000  NaN         NaN  75.850000  NaN  NaN  NaN
75%     7.000000  NaN         NaN  76.150000  NaN  NaN  NaN
max     9.000000  NaN         NaN  76.400000  NaN  NaN  NaN

## 인덱싱 (색인)
- 넘파이의 배열 인덱스와 유사
- 차이점은 **배열은 행우선 인덱스**
- 판다스의 데이터프레임은 기본적으로 **열우선** 인덱스
    - **주의** 데이터프레임에서는 행을 검색하려고 하면 안 됨

### 열 인덱싱

In [ ]:
#컬럼이름이 컬럼인덱스?

In [ ]:
# 열을 기준으로 시리즈를 선택할 수 있음
# 하나만 있는 경우 열인덱스를 의미하는 것
# 데이터 프레임은 행을 선택하지 않고, 시리즈를 선택하는 방향으로 인덱스 함
rawData['몸무게']


0    76.4
1    75.7
2    76.0
3     NaN
4    76.2
5    75.7
6     NaN
7     NaN
8    75.0
Name: 몸무게, dtype: float64

In [ ]:
type( rawData['몸무게'] )

pandas.core.series.Series

In [ ]:
# 배열 인덱스도 지원 함
col = ['이름', '몸무게','측정일']
rawData[col]

이름   몸무게         측정일
0  홍길동  76.4  2020-03-01
1  홍길동  75.7  2020-03-02
2  홍길동  76.0  2020-03-03
3  홍길동   NaN  2020-03-04
4  홍길동  76.2  2020-03-05
5  홍길동  75.7  2020-03-06
6  홍길동   NaN  2020-03-07
7  홍길동   NaN  2020-03-08
8  홍길동  75.0  2020-03-09

### 행 인덱싱
- 당연히 행도 선택할 수 있어야 함
- loc, iloc 을 이용해서
- 해당 행으 **시리즈 형태**로 반환해줌


In [ ]:
#첫번째 행을 선택해보자
rawData.head(1)

회차   이름         측정일   몸무게  단위   담당   지점
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구

In [ ]:
# 위의 0번 인덱스를 선택하려면
rawData.loc[0]

회차              1
이름            홍길동
측정일    2020-03-01
몸무게          76.4
단위             kg
담당            박현경
지점            관악구
Name: 0, dtype: object

In [ ]:
# loc 와 iloc 는 아직까지는 별 차이가 없음. 차이는 추후에 다룰 듯?
type(rawData.loc[0])

pandas.core.series.Series

In [ ]:
rawData.iloc[0]

회차              1
이름            홍길동
측정일    2020-03-01
몸무게          76.4
단위             kg
담당            박현경
지점            관악구
Name: 0, dtype: object

In [ ]:
# 뒤집기.
# 행과 열을 바꿈 (Transfer?)
rawData.T

0           1           2  ...           6           7           8
회차            1           2           3  ...           7           8           9
이름          홍길동         홍길동         홍길동  ...         홍길동         홍길동         홍길동
측정일  2020-03-01  2020-03-02  2020-03-03  ...  2020-03-07  2020-03-08  2020-03-09
몸무게        76.4        75.7          76  ...         NaN         NaN          75
단위           kg          kg          kg  ...          kg          kg          kg
담당          박현경         김현경         최현경  ...         최현경         김현경         김현경
지점          관악구         관악구         여의도  ...         서초구         서초구         서초구

[7 rows x 9 columns]

### 슬라이스도 가능
- 판다스의 슬라이스는 **마지막 인덱스 포함**하고 있다


In [ ]:
rawData.loc[0:3]
# 마지막 인덱스 3까지 포함

회차   이름         측정일   몸무게  단위   담당   지점
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도

In [ ]:
# 행에 대해서 배열 인덱싱
rawData.loc[ [1, 3, 5] ]

회차   이름         측정일   몸무게  단위   담당   지점
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구

In [ ]:
# 보통은 location 사용하긴 함
rawData.reindex( [1, 3, 5] )

회차   이름         측정일   몸무게  단위   담당   지점
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구

### 행과 열을 같이 인덱싱

In [ ]:
rawData.loc[0, '몸무게']

76.4

In [ ]:
# 물론 배열 인덱스와 함께 사용할 수 있다
rawData.loc[0, ['이름', '몸무게'] ]

이름      홍길동
몸무게    76.4
Name: 0, dtype: object

In [ ]:
# 행도 가능
rawData.loc[[1,3,5], ['이름', '몸무게'] ]

이름   몸무게
1  홍길동  75.7
3  홍길동   NaN
5  홍길동  75.7

In [ ]:
# 슬라이스와 함께 사용
rawData.loc[0:3, ['이름', '몸무게']]

이름   몸무게
0  홍길동  76.4
1  홍길동  75.7
2  홍길동  76.0
3  홍길동   NaN

In [ ]:
rawData.loc[0:3, '이름':'몸무게']

이름         측정일   몸무게
0  홍길동  2020-03-01  76.4
1  홍길동  2020-03-02  75.7
2  홍길동  2020-03-03  76.0
3  홍길동  2020-03-04   NaN

### loc vs. iloc
- iloc를 사용하면 컬럼(행) 인덱스에 정수를 사용할 수 있다.


In [ ]:
rawData.iloc[0:3, 1:4]

이름         측정일   몸무게
0  홍길동  2020-03-01  76.4
1  홍길동  2020-03-02  75.7
2  홍길동  2020-03-03  76.0

## 조건검색
- 불리언 인덱스의 활용


In [ ]:
rawData

회차   이름         측정일   몸무게  단위   담당   지점
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구

In [ ]:
# 지점별로 여의도 자료만 확인하고 싶다면
rawData['지점'] == '여의도'
# 불리언 시리즈가 반환된다.
# 생성된 이 결과를 인덱스로 사용

0    False
1    False
2     True
3     True
4    False
5    False
6    False
7    False
8    False
Name: 지점, dtype: bool

In [ ]:
rawData[ rawData['지점'] == '여의도' ]

회차   이름         측정일   몸무게  단위   담당   지점
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도

In [ ]:
rawData[ rawData['담당'] == '김현경' ]

회차   이름         측정일   몸무게  단위   담당   지점
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구

In [ ]:
# 숫자도 가능
# 몸무게가 76 이상인 자료만 검색하려면
rawData[ rawData['몸무게'] >= 76 ]

회차   이름         측정일   몸무게  단위   담당   지점
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구

### 다중 조건
- 파이썬의 and, or, not 사용처럼
- 판다스는 &, |, ~ 문자를 사용해서 표현
  -&(앰퍼센드), |(파이프라인),~(틸드)
  - 우선순위가 매우 헷갈림
  - 그렇기 때문에 괄호를 이용해서 우선순위를 정확하게 표현하는 것이 좋다


In [ ]:
# 서초구 지점의 최현경 담담의 자료만 검색하려면
rawData[ rawData['지점'] == '서초구' & rawData['담당'] == '최현경']

TypeError: ignored

- 이렇게 에러가 나거나 원하는 대로 처리가 안 될 수 있음
- 괄호로 구분지어주는게 좋음

In [ ]:
# 서초구 지점의 최현경 담담의 자료만 검색할 때 괄호를 치면
rawData[ (rawData['지점'] == '서초구') & (rawData['담당'] == '최현경') ]

회차   이름         측정일   몸무게  단위   담당   지점
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구

In [ ]:
# 서초구 지점의 김현경과 최현경 담당자의 자료만 검색하려면
rawData[ (rawData['지점'] == '서초구' ) & ( (rawData['담당'] == '김현경') | (rawData['담당'] == '최현경') )]

회차   이름         측정일   몸무게  단위   담당   지점
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구

In [ ]:
# 지점이 서초구면서, 담장자가 최현경이 아닌 자료를 검색하려면
rawData[ (rawData['지점'] == '서초구' ) & ~ (rawData['담당'] == '최현경') ]

회차   이름         측정일   몸무게  단위   담당   지점
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구

### 문자열 처리

In [ ]:
# isin은 리스트를 파라미터로 갖는다
# 리스트 내의 값들 중에서 하나라도 존재하면 True, 그렇지 않으면 False
rawData[ (rawData['지점'] == '서초구') & (rawData['담당'].isin(['최현경', '김현경']))]

회차   이름         측정일   몸무게  단위   담당   지점
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구

In [ ]:
# 담당자들 중에 성이 '김'씨인 자료만 검색
# 파이썬의 문자열 메소드 이용하는 경우
# str을 사용하면 파이썬에서 제공하는 문자열 메소드를 동일하게 사용할 수 있다
# coantains는 문자열 내에서 특정 문자열이 존재하면 True, 그렇지 않으면 False
rawData[ rawData['담당'].str.contains('김') ] 


회차   이름         측정일   몸무게  단위   담당   지점
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구

In [ ]:
# statrwith 는 문자열 내에서 특정 문자열로 시작하는 경우 True, 그렇지 않으면 False
rawData[ rawData['담당'].str.startswith('김')]

회차   이름         측정일   몸무게  단위   담당   지점
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구

### 결측치
- 결측치가 있으면 분석을 제대로 수행할 수없기 때문에 어떻게든 처리를 해줘야한다.
  - 결측치를 채우든가, 지우든가
  - 결측치를 채우는 방법
    - 결측치를 다른 값으로 대체하는 방법
      - 평균, 중앙값 등으로
      - (강사님은 개인적으로 별로..)
    - 

In [ ]:
# 결측치를 확인하는 방법
# 결측치의 개수
rawData['몸무게'].isnull()

0    False
1    False
2    False
3     True
4    False
5    False
6     True
7     True
8    False
Name: 몸무게, dtype: bool

In [ ]:
rawData['몸무게'].isna()

0    False
1    False
2    False
3     True
4    False
5    False
6     True
7     True
8    False
Name: 몸무게, dtype: bool

In [ ]:
# True는 1이기 때문에 다 더하면 True값을 모두 더하면 그게 개수가 됨
rawData['몸무게'].isna().sum()


3

In [ ]:
# 결측치만 확인하고 싶은 경우
rawData[rawData['몸무게'].isna() ]

회차   이름         측정일  몸무게  단위   담당   지점
3   4  홍길동  2020-03-04  NaN  kg  최현경  여의도
6   7  홍길동  2020-03-07  NaN  kg  최현경  서초구
7   8  홍길동  2020-03-08  NaN  kg  김현경  서초구

In [ ]:
# 결측치를 제외하고 확인하고 싶은 경우
# 부정을 해도 되고
rawData[ ~ rawData['몸무게'].isna() ]

회차   이름         측정일   몸무게  단위   담당   지점
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구

In [ ]:
# 메소드를 사용해도 되고
rawData[rawData['몸무게'].notna() ]

회차   이름         측정일   몸무게  단위   담당   지점
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구

In [ ]:
# 조건에 맞는 자료등 중에서 내가 원하는 컬럼만 확인하려면
# 결측치인 경우의 이름만 확인하는 법
# 로케이트를 같이 써줘야 함
rawData.loc[rawData['몸무게'].isna(), '이름' ]

3    홍길동
6    홍길동
7    홍길동
Name: 이름, dtype: object

In [ ]:
# 몸무게를 측정하지 않은 자료들의 지점, 담당, 측정일을 확인하려면
rawData.loc[rawData['몸무게'].isna(), ['지점', '담당', '측정일'] ]

지점   담당         측정일
3  여의도  최현경  2020-03-04
6  서초구  최현경  2020-03-07
7  서초구  김현경  2020-03-08

### 이상치 (outlier)
- 특별히 크거나, 작은 값
  - 얼마나 크거나 작아야 이상치라고 할  수있을까?
    - 이건 분석자가 정하든..?
- 이상치를 찾는 경우가 아니라면, 이상치는 제거하고 분석을 진행한다.

- 자료의 분포를 봤을 때 이건 너무 크구나 싶은 경우엔 이상치를 제거한다는 듯?

In [ ]:
rawData['몸무게'].describe()

count     6.000000
mean     75.833333
std       0.492612
min      75.000000
25%      75.700000
50%      75.850000
75%      76.150000
max      76.400000
Name: 몸무게, dtype: float64

In [ ]:
# 위의 값이 이상치가 있진 않지만 억지로 이상치를 설정해본다면,
# 76을 넘어가면 이상치로 판단한다고 가정한다면
rawData[ rawData['몸무게'] > 76.15 ]

회차   이름         측정일   몸무게  단위   담당   지점
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구

In [ ]:
# 4분위수를 이용한 방법
rawData['몸무게'].sort_values()
# 분포도에서 중간의 50%, 상위/하위 25%씩, 그리고 최대값/최소값 1% 4개로 나눈건가
# 자세한 건 추후에 다룰 거임...

8    75.0
1    75.7
5    75.7
2    76.0
4    76.2
0    76.4
3     NaN
6     NaN
7     NaN
Name: 몸무게, dtype: float64

In [ ]:
display(rawData['몸무게'].quantile(0.25))
display(rawData['몸무게'].quantile(0.50))
display(rawData['몸무게'].quantile(0.75))
display(rawData['몸무게'].quantile(0.99))


75.7

75.85

76.15

76.39

In [ ]:
# 이상치를 상위 99%를 넘어가는 값이라고 지정한다면
# 하위 0.1%보다 낮은 값도 이상치로 지정한다면
# 이상치는 제거하는 방법
low = rawData['몸무게'].quantile(0.01)
high = rawData['몸무게'].quantile(0.99)
rawData[ (rawData['몸무게'] < high) & (rawData['몸무게'] > low) ]

회차   이름         측정일   몸무게  단위   담당   지점
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구

### 중복 데이터 검사
- duplicated 를 이용해서 중복된 자료를 검사
  - 중복된 자료를 삭제하거나 합쳐주거나

In [ ]:
rawData.duplicated(subset=['지점'])

0    False
1     True
2    False
3     True
4    False
5    False
6     True
7     True
8     True
dtype: bool

In [ ]:
rawData[ rawData.duplicated(subset=['지점'])]
# 중복되는 값들 중 제일 먼저의 값 말고 그 다음부터 중복되는 값들을 보여준 듯!

회차   이름         측정일   몸무게  단위   담당   지점
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구

In [ ]:
# keep=False 로 파라미터를 지정해주면
# 모든 중복된 데이터를 보여줌
rawData[ rawData.duplicated(subset=['지점'], keep=False) ]

회차   이름         측정일   몸무게  단위   담당   지점
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구

##통계분석
- 판다스에서 제공하는 통게관련된 기능들
  - pivot
    - (엑셀에서 사용하는 피봇 기능과 동일)(피벗 말하는 듯?..)
  - crosstab 
    - 변수가 두개 이상?
  - 그룹화

In [ ]:
# 지점별 몸무게 평균
pd.pivot_table( rawData, index='지점', values='몸무게' )
# 기본값이 평균을 구해주는 거임

몸무게
지점        
강남구  76.20
관악구  76.05
서초구  75.35
여의도  76.00

In [ ]:
pd.pivot_table( rawData, index='담당', values='몸무게', aggfunc=np.sum )

몸무게
담당        
김현경  226.9
박현경  153.4
최현경  151.7

In [ ]:
# 지점별 몸무게 총합
pd.pivot_table( rawData, index='지점', values='몸무게', aggfunc=np.sum )

몸무게
지점        
강남구   76.2
관악구  152.1
서초구  150.7
여의도   76.0

In [ ]:
# 지점별 담당자별 몸무게의 평균
# 인덱스 부분을 리스트로 만들어줌..?
pd.pivot_table( rawData, index=['지점', '담당'], values='몸무게', aggfunc=np.sum )

몸무게
지점  담당       
강남구 김현경  76.2
관악구 김현경  75.7
    박현경  76.4
서초구 김현경  75.0
    최현경  75.7
여의도 최현경  76.0

In [ ]:
# 지점별 담당자별 평균
pd.crosstab( index=rawData['지점'], columns=rawData['담당'], values=rawData['몸무게'], aggfunc=np.mean )

담당    김현경   박현경   최현경
지점                   
강남구  76.2   NaN   NaN
관악구  75.7  76.4   NaN
서초구  75.0   NaN  75.7
여의도   NaN   NaN  76.0

In [ ]:
# 그룹화를 통한 통계 수치
# 그룹 객체를 반환
rawData.groupby(['지점'])

In [ ]:
# 그룹화를 통한 지점별 몸무게 평균
rawData.groupby(['지점'])['몸무게'].mean()


지점
강남구    76.20
관악구    76.05
서초구    75.35
여의도    76.00
Name: 몸무게, dtype: float64

In [ ]:
# 지점별 담당자별 몸무게 평균
rawData.groupby(['지점', '담당'])[['몸무게']].mean()
# 몸무게가 1차원이어도 [[]] 처럼 2차원으로 표현한다면 프레임 할 수 ㅣㅆ다?

몸무게
지점  담당       
강남구 김현경  76.2
관악구 김현경  75.7
    박현경  76.4
서초구 김현경  75.0
    최현경  75.7
여의도 최현경  76.0

In [ ]:
# 컬럼별로 집계를 다르게 하고 싶다면
# 지점별 몸무게 평균과 담당의 수
# count가 없어서 문자열로 처리했는데 되네?...
# np.mean 도 'mean'으로 해도 된다고 함...왜지?
rawData.groupby('지점').agg({
    '몸무게':np.mean, '담당': 'count'
})

몸무게  담당
지점            
강남구  76.20   1
관악구  76.05   2
서초구  75.35   4
여의도  76.00   2

# 데이터 프ㅔ임 조작하기
- 추가, 수정 ,삭제

## 컬럼의 추가와 수정
- 컬럼을 추가하거나 수정하는데 큰 기능이 필요하진 않다.
- 그냥 쓰면 됨?

In [ ]:
# 데이터 프레임에 컬럼이 존재하지 않으면 새로 생성
# 지역이라는 컬럼을 추가해보자
rawData['지역'] = '서울'
rawData

회차   이름         측정일   몸무게  단위   담당   지점  지역
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구  서울
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구  서울
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도  서울
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도  서울
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구  서울
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구  서울
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구  서울
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구  서울
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구  서울

In [ ]:
# 해당 컬럼이 존재하면 수정이 된다
rawData['지역'] = 'Soeul'
rawData

회차   이름         측정일   몸무게  단위   담당   지점     지역
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구  Soeul
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구  Soeul
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도  Soeul
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도  Soeul
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구  Soeul
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구  Soeul
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구  Soeul
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구  Soeul
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구  Soeul

In [ ]:
# 조건에 맞춰서 컬럼을 만들어 보지
# 몸무게가 특정값 이상이면 '비만', 그렇지 않으면 '정상'을 갖는 '상태' 컬럼을 추가
# 조건검색과 함께 사용
rawData.loc[ rawData['몸무게'] >= 76, '상태'] = '비만'
rawData

회차   이름         측정일   몸무게  단위   담당   지점     지역   상태
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구  Soeul   비만
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구  Soeul  NaN
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도  Soeul   비만
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도  Soeul  NaN
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구  Soeul   비만
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구  Soeul  NaN
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구  Soeul  NaN
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구  Soeul  NaN
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구  Soeul  NaN

In [ ]:
rawData.loc[ rawData['몸무게'] < 76, '상태' ] = '정상'
rawData

회차   이름         측정일   몸무게  단위   담당   지점     지역   상태
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구  Soeul   비만
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구  Soeul   정상
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도  Soeul   비만
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도  Soeul  NaN
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구  Soeul   비만
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구  Soeul   정상
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구  Soeul  NaN
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구  Soeul  NaN
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구  Soeul   정상

In [ ]:
#########몸무게가 float 상태가 아니거나 그런 듯 일단 넘어가고 이따 다시 확인해보자
부등호는 숫자랑 숫자 비교해야되니까

## 컬럼삭제
- 삭제는 신중하게 진행해야 한다
  - 귀찮아도 원본 데이터는 삭제하지 않는 것이 좋다

In [ ]:
# drop를 이용해 특정 컬럼을 삭제할 수 있다.
# drop를 사용하는 경우 특정 컬럼이 삭제된 새로운 데이터프레임을 반환
rawData.drop(columns=['지역'])

회차   이름         측정일   몸무게  단위   담당   지점   상태
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구   비만
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구   정상
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도   비만
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도  NaN
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구   비만
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구   정상
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구  NaN
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구  NaN
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구   정상

In [ ]:
# 원본은 바뀌지 않는다
rawData

회차   이름         측정일   몸무게  단위   담당   지점     지역   상태
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구  Soeul   비만
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구  Soeul   정상
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도  Soeul   비만
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도  Soeul  NaN
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구  Soeul   비만
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구  Soeul   정상
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구  Soeul  NaN
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구  Soeul  NaN
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구  Soeul   정상

In [ ]:
# 원본에서 바로 삭제할 수 있다
# 따로 반환되지 않는다
rawData.drop( columns=['지역'], inplace=True)

In [ ]:
# inplace 파라미터를 True로 하면, 해당 데이터프레임에서 바로 삭제
rawData

회차   이름         측정일   몸무게  단위   담당   지점   상태
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구   비만
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구   정상
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도   비만
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도  NaN
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구   비만
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구   정상
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구  NaN
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구  NaN
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구   정상

## 행 추가와 수정, 삭제

In [ ]:
# 행이 존재하지 않으면 그냥 추가가 됨
rawData
# 행인덱스가 0부터 8번까지 있네

회차   이름         측정일   몸무게  단위   담당   지점   상태
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구   비만
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구   정상
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도   비만
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도  NaN
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구   비만
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구   정상
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구  NaN
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구  NaN
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구   정상

In [ ]:
rawData.loc[9]=[10, '홍길동', '2020-03-10', 77, 'kg', '박현경', '관악구', '비만',]
rawData
# 행이 추가 됨

회차   이름         측정일   몸무게  단위   담당   지점   상태
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구   비만
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구   정상
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도   비만
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도  NaN
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구   비만
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구   정상
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구  NaN
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구  NaN
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구   정상
9  10  홍길동  2020-03-10  77.0  kg  박현경  관악구   비만

In [ ]:
# 삭제는 컬럼과 마찬가지로 drop을 동일하게 사용
# 대신 파라미터가 columns 대신 index 파라미터를 사용
rawData.drop(index=[0, 1, 2])
# [0], [1], [2] 의 행 삭제됨

회차   이름         측정일   몸무게  단위   담당   지점   상태
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도  NaN
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구   비만
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구   정상
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구  NaN
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구  NaN
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구   정상
9  10  홍길동  2020-03-10  77.0  kg  박현경  관악구   비만

## apply
  -  (많이 쓰이는 기능임)
  - 자료를 내가 원하는 대로 다루기는 쉽지 않음
  -실제로 자료를 다뤄보면 내 생각처러 쉽게 다뤄지지 않음. 상당히 어려움
  - 자료의 형태가 천차만별이고, 지저분한 경우도 많음
- 판다스에서 사용가능한 반복문 정도로 이해
- **모든 행, 열에 대해서 동일한 함수를 반복적으로 적용할 수 있다.**


In [ ]:
# apply의 기본적인 사용방법
def func(x):
  print('함수가 호출되었습니다')
  print('x: {}'.format(x) )

# 기본적으로는 데이터 프레임의 각 컬럼들이 시리즈 형태로 전달된다고 보면 됨
  return
#apply를 통해서 호출되는 함수는 정의할 때 return을 꼭 넣어줄 것
rawData.apply( func )
# 마지막에 회차가 넘어갔을 때 리턴값이 없어서 None이 나오는 것이라고 함
# 이따 확인.. 일단 넘어감

함수가 호출되었습니다
x: 0     1
1     2
2     3
3     4
4     5
5     6
6     7
7     8
8     9
9    10
Name: 회차, dtype: int64
함수가 호출되었습니다
x: 0    홍길동
1    홍길동
2    홍길동
3    홍길동
4    홍길동
5    홍길동
6    홍길동
7    홍길동
8    홍길동
9    홍길동
Name: 이름, dtype: object
함수가 호출되었습니다
x: 0    2020-03-01
1    2020-03-02
2    2020-03-03
3    2020-03-04
4    2020-03-05
5    2020-03-06
6    2020-03-07
7    2020-03-08
8    2020-03-09
9    2020-03-10
Name: 측정일, dtype: object
함수가 호출되었습니다
x: 0    76.4
1    75.7
2    76.0
3     NaN
4    76.2
5    75.7
6     NaN
7     NaN
8    75.0
9    77.0
Name: 몸무게, dtype: float64
함수가 호출되었습니다
x: 0    kg
1    kg
2    kg
3    kg
4    kg
5    kg
6    kg
7    kg
8    kg
9    kg
Name: 단위, dtype: object
함수가 호출되었습니다
x: 0    박현경
1    김현경
2    최현경
3    최현경
4    김현경
5    최현경
6    최현경
7    김현경
8    김현경
9    박현경
Name: 담당, dtype: object
함수가 호출되었습니다
x: 0    관악구
1    관악구
2    여의도
3    여의도
4    강남구
5    서초구
6    서초구
7    서초구
8    서초구
9    관악구
Name: 지점, dtype: object
함수가 호출되었습니다
x: 0     비만
1     정상
2 

회차     None
이름     None
측정일    None
몸무게    None
단위     None
담당     None
지점     None
상태     None
dtype: object

In [ ]:
# 컬럼이 아니라 행을 전달하고 싶다면
def func(x):
  print('함수가 호출되었습니다')
  print('x: {}'.format(x) )

# 디폴트 값이 컬럼으로 되어 있어서 axis 파라미터에 지정하면 됨
# 0이면 컬럼, 1이면 행
rawData.apply( func, axis=1 )

함수가 호출되었습니다
x: 회차              1
이름            홍길동
측정일    2020-03-01
몸무게          76.4
단위             kg
담당            박현경
지점            관악구
상태             비만
Name: 0, dtype: object
함수가 호출되었습니다
x: 회차              2
이름            홍길동
측정일    2020-03-02
몸무게          75.7
단위             kg
담당            김현경
지점            관악구
상태             정상
Name: 1, dtype: object
함수가 호출되었습니다
x: 회차              3
이름            홍길동
측정일    2020-03-03
몸무게            76
단위             kg
담당            최현경
지점            여의도
상태             비만
Name: 2, dtype: object
함수가 호출되었습니다
x: 회차              4
이름            홍길동
측정일    2020-03-04
몸무게           NaN
단위             kg
담당            최현경
지점            여의도
상태            NaN
Name: 3, dtype: object
함수가 호출되었습니다
x: 회차              5
이름            홍길동
측정일    2020-03-05
몸무게          76.2
단위             kg
담당            김현경
지점            강남구
상태             비만
Name: 4, dtype: object
함수가 호출되었습니다
x: 회차              6
이름            홍길동
측정일    2020-03-06
몸무게          75.7
단위 

0    None
1    None
2    None
3    None
4    None
5    None
6    None
7    None
8    None
9    None
dtype: object

In [ ]:
# 지점별로 지역을 새로 만들어 보려면
def func(x):
  if x['지점'] == '관악구': return '강서'
  elif x['지점'] == '여의도': return '강서'
  elif x['지점'] == '강남구' : return '강동'
  elif x['지점'] == '서초구' : return '강동'

rawData.apply( func, axis=1 )
# apply가 리턴해주는 값도 시리즈라서 바로 프레임에 추가할 수도 있음
# 그래서 이 시리즈에도 인덱스가 붙음
# 해당 인덱스에 컬럼을 추가할 수도 있음

0    강서
1    강서
2    강서
3    강서
4    강동
5    강동
6    강동
7    강동
8    강동
9    강서
dtype: object

In [ ]:
# 리턴해준 인덱스에 새로운 컬럼을 추가
rawData['지역'] = rawData.apply( func, axis=1 )
rawData

회차   이름         측정일   몸무게  단위   담당   지점   상태  지역
0   1  홍길동  2020-03-01  76.4  kg  박현경  관악구   비만  강서
1   2  홍길동  2020-03-02  75.7  kg  김현경  관악구   정상  강서
2   3  홍길동  2020-03-03  76.0  kg  최현경  여의도   비만  강서
3   4  홍길동  2020-03-04   NaN  kg  최현경  여의도  NaN  강서
4   5  홍길동  2020-03-05  76.2  kg  김현경  강남구   비만  강동
5   6  홍길동  2020-03-06  75.7  kg  최현경  서초구   정상  강동
6   7  홍길동  2020-03-07   NaN  kg  최현경  서초구  NaN  강동
7   8  홍길동  2020-03-08   NaN  kg  김현경  서초구  NaN  강동
8   9  홍길동  2020-03-09  75.0  kg  김현경  서초구   정상  강동
9  10  홍길동  2020-03-10  77.0  kg  박현경  관악구   비만  강서

# 데이터프레임 합치기
- 서로 다른 두 데이터프레임이 있을 때,
- 하나의 데이터 프레임으로 합치는 방법
- 여러 소스(출처)로부터 가져온 데이터를 분석을 하기위해서는 반드시 하나의 데이터 프레임이어야만 한다
- 데이터를 합치는 방법

![](https://img1.daumcdn.net/thumb/R1280x0/?scode=mtistory2&fname=https%3A%2F%2Fblog.kakaocdn.net%2Fdn%2Foecbm%2FbtqRolfgnC4%2FROZxkKlASUIqWLx1SLd7xK%2Fimg.png)

In [ ]:
user_device = pd.read_csv('/content/drive/MyDrive/멀티캠퍼스/Data폴더/user_device.csv')

In [ ]:
user_usage = pd.read_csv('/content/drive/MyDrive/멀티캠퍼스/Data폴더/user_usage.csv')

In [ ]:
user_usage.head(2)

outgoing_mins_per_month  outgoing_sms_per_month  monthly_mb  use_id
0                    21.97                    4.82     1557.33   22787
1                  1710.08                  136.88     7267.55   22788

In [ ]:
user_device.head(2)

use_id  user_id platform  platform_version     device  use_type_id
0   22782    26980      ios              10.2  iPhone7,2            2
1   22783    29628  android               6.0    Nexus 5            3

In [ ]:
display( len(user_usage) )
display( len(user_device) )

240

272

## inner join
- merge 함수를 통해서 `join`을 해볼 수 있습니다. 
  - merge의 기본 동작은 `inner join` 입니다. 

- inner join은 기본적으로 자료의 크기가 훨씬 줄어들게 됩니다. 
  - user_usage: 240개
  - user_device: 272개
    - 두 자료를 하나로 합쳤을 경우 240개를 넘어가지는 않습니다. 
    - 양쪽자료 모두에 존재하지 않는 자료가 있다면? 

In [ ]:
display( len(user_usage) )
display( len(user_device) )

240

272

In [ ]:
pd.merge(left=user_usage, right=user_device, on='use_id')

outgoing_mins_per_month  ...  use_type_id
0                      21.97  ...            1
1                    1710.08  ...            1
2                    1710.08  ...            1
3                      94.46  ...            1
4                      71.59  ...            1
..                       ...  ...          ...
154                   198.59  ...            1
155                   198.59  ...            1
156                   106.65  ...            1
157                   344.53  ...            1
158                    42.75  ...            1

[159 rows x 9 columns]

## left join
- left 자료를 기준으로 합쳐줍니다. 
- 결측치가 발생할 수 있습니다. 
  - right에 없는 자료는 결측치로 채워지게 됩니다. 

In [ ]:
# how의 기본 파라미터는 inner join
pd.merge(left=user_usage, right=user_device, on='use_id', how='left')

outgoing_mins_per_month  outgoing_sms_per_month  ...    device  use_type_id
0                      21.97                    4.82  ...  GT-I9505          1.0
1                    1710.08                  136.88  ...  SM-G930F          1.0
2                    1710.08                  136.88  ...  SM-G930F          1.0
3                      94.46                   35.17  ...     D2303          1.0
4                      71.59                   79.26  ...  SM-G361F          1.0
..                       ...                     ...  ...       ...          ...
235                   260.66                   68.44  ...       NaN          NaN
236                    97.12                   36.50  ...       NaN          NaN
237                   355.93                   12.37  ...       NaN          NaN
238                   632.06                  120.46  ...       NaN          NaN
239                   488.70                  906.92  ...       NaN          NaN

[240 rows x 9 columns]

## full outer join

In [ ]:
pd.merge(left=user_usage, right=user_device, on='use_id', how='outer')

outgoing_mins_per_month  ...  use_type_id
0                      21.97  ...          1.0
1                    1710.08  ...          1.0
2                    1710.08  ...          1.0
3                      94.46  ...          1.0
4                      71.59  ...          1.0
..                       ...  ...          ...
348                      NaN  ...          2.0
349                      NaN  ...          3.0
350                      NaN  ...          3.0
351                      NaN  ...          3.0
352                      NaN  ...          3.0

[353 rows x 9 columns]

# 결론
- 데이터 전처리 작업의 핵심은 합치고, 나누고, 붙이고, 자르는 작업
  - 무한히 반복
  - `인공지능`이라는 개념이 나오면서 더더욱 오해가 심해져 가고 있는데
    - 뭔가 업무가 스마트할 것 같고, 있어 보이는 이런 작업
    - 실제 작업은 그렇게 스마트 하지 않다